core/crypto/sm3.c

```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright Copyright (c) 2019 Huawei Technologies Co., Ltd
+ */
+/*
+ * SM3 Hash algorithm
+ * thanks to Xyssl
+ * author:goldboar
+ * email:goldboar@163.com
+ * 2011-10-26
+ */
+
+#include <io.h>
+#include <string.h>
+#include <string_ext.h>
+
+#include "sm3.h"
+
+void sm3_init(struct sm3_context *ctx)
+{
+	ctx->total[0] = 0;
+	ctx->total[1] = 0;
+
+	ctx->state[0] = 0x7380166F;
+	ctx->state[1] = 0x4914B2B9;
+	ctx->state[2] = 0x172442D7;
+	ctx->state[3] = 0xDA8A0600;
+	ctx->state[4] = 0xA96F30BC;
+	ctx->state[5] = 0x163138AA;
+	ctx->state[6] = 0xE38DEE4D;
+	ctx->state[7] = 0xB0FB0E4E;
+}
+
+static void sm3_process(struct sm3_context *ctx, const uint8_t data[64])
+{
+	uint32_t SS1, SS2, TT1, TT2, W[68], W1[64];
+	uint32_t A, B, C, D, E, F, G, H;
+	uint32_t T[64];
+	uint32_t Temp1, Temp2, Temp3, Temp4, Temp5;
+	int j;
+
+	for (j = 0; j < 16; j++)
+		T[j] = 0x79CC4519;
+	for (j = 16; j < 64; j++)
+		T[j] = 0x7A879D8A;
+
+	W[0] = get_be32(data);
```

Comment:
We're assuming that unaligned 32-bit access is OK here.

Reply:
Fixed